### PR TITLE
feat(features): Add a default manager

### PR DESCRIPTION
**Pull request by wmak**
- Adding an optional entity manager, which when included will be the checked after registered flags, and before chceking settings
- Adds an `organizations:trend` feature to test the functionality of the `entity_manager`
- This allows the entity handler introduced in getsentry/getsentry#4362 to be used when we check features with `features.has` and to make migration of flags easy

### CODE REVIEW

File: src/sentry/features/__init__.py [hunk 1]
```diff
@@ -127,10 +127,13 @@
     "organizations:incidents",
 )
 
+default_manager.add("organizations:trends", OrganizationFeature)  # NOQA
```
Comment (markstory) [inline src/sentry/features/__init__.py:130]: Should this be performance-trends to be more specific?

File: src/sentry/features/manager.py [hunk 2]
```diff
@@ -49,6 +51,15 @@ def get(self, name, *args, **kwargs):
         cls = self._get_feature_class(name)
         return cls(name, *args, **kwargs)
 
+    def add_default_handler(self, handler, backup_handler=None):
```
Comment (markstory) [inline src/sentry/features/manager.py:54]: Would a name like 'entity type' handler be applicable here? Having default and backup in the same class is a bit confusing to me.

File: src/sentry/features/manager.py [hunk 3]
```diff
@@ -85,11 +96,21 @@ def has(self, name, *args, **kwargs):
...
+            rv = self._default_handler(feature, actor)
+            if rv is not None:
+                return rv
+
         # Check registered feature handlers
         rv = self._get_handler(feature, actor)
         if rv is not None:
             return rv
 
+        if self._backup_handler:
```
Comment (markstory) [inline src/sentry/features/manager.py:109]: Would fallback work as a term here?

**Review by markstory — COMMENTED** [review]
Should we have tests to ensure that entity type/normal/fallback flow works?

File: src/sentry/features/manager.py [hunk 4]
```diff
@@ -85,11 +96,21 @@ def has(self, name, *args, **kwargs):
         actor = kwargs.pop("actor", None)
         feature = self.get(name, *args, **kwargs)
 
+        if self._default_handler:
```
Comment (evanh) [inline src/sentry/features/manager.py:99]: I'm a little confused by this. Default normally implies the same thing as backup. "Use this value if no other value is set". This seems to be an override more than anything.

**Review by evanh — APPROVED** [review]

File: src/sentry/features/manager.py [hunk 5]
```diff
@@ -90,6 +97,11 @@ def has(self, name, *args, **kwargs):
         if rv is not None:
             return rv
 
+        if self._entity_handler:
```
Comment (markstory) [inline src/sentry/features/manager.py:100]: Should the docstring for this method be updated to reflect that the entity handler will also be checked?

**Review by markstory — APPROVED** [review]

**Comment by dcramer** [comment]
left a related comment in getsentry/getsentry#4362 but i dont understand what the purpose of this change is. at the very least we should clearly define it, because i quite honestly cant comprehend what it 1) means, 2) is intended to solve, by looking at the code or this PR

in general, id ask ourselves if we really need this extra complexity. the feature flagging api has gotten awfully complicated from its original design, and no ones done anything to reduce that complexity, only add to it. if we really need different behavior, its worth considering if we can just swap out APIs instead of adding more branching. we dont actually need the level of extensibility i originally created (theres really not a serious need for more than one feature handler, as an example, and no one extends this system besides us these days)

**Comment by wmak** [comment]
@dcramer 
So there's a few reasons I think we should stick with the current module for now, and remove complexity once we no longer need to depend on the multitude of existing handlers.
- While we're testing the new system out it'd be nice to keep the current API, and have the fallback flow from: Existing Handlers > Entity Handler > Settings
- `has_for_batch` can take advantage of bulk evaluation with the Entity Handler

But to make sure I'm not missing the point here, the complexity mentioned is around registering new features right? ie.
- Adding it [here](https://github.com/getsentry/sentry/pull/20340/files#diff-75828aa0bdaf3e531c6672d79b569845R130)
- Having to check feature handlers [here](https://github.com/getsentry/sentry/pull/20340/files#diff-39fdf09adfa696684df2dd28968285e0R112) to handle each flag

**Comment by dcramer** [comment]
@wmak its more "how do features work" - i have to parse some quite complicated branches of logic that aren't really explained well. they look like internal apis more than they do something that is production stable. At the very least I'd document the intent long term (in code, comments), and think about the naming here so its more obvious whats going on.

**Pull request merged**